keywrap/pkcs11/keywrapper_pkcs11.go

```diff
@@ -0,0 +1,130 @@
+func NewKeyWrapper() keywrap.KeyWrapper {
+	return &pkcs11KeyWrapper{}
+}
+
+// WrapKeys wraps the session key for recpients and encrypts the optsData, which
+// describe the symmetric key used for encrypting the layer
+func (kw *pkcs11KeyWrapper) WrapKeys(ec *config.EncryptConfig, optsData []byte) ([]byte, error) {
+	// FIXME: The pkcs11-uris may be insufficient for more advanced pkcs11 modules that need
+	// environment variables set. So, public keys may have to be 'pkcs11-yamls' as well, so
+	// that we can set environment variables for the pkcs11 module to find config file etc.
+	pkcs11Recipients, err := addPubKeys(&ec.DecryptConfig, append(ec.Parameters["pkcs11-pubkeys"], ec.Parameters["pkcs11-uris"]...))
+	if err != nil {
+		return nil, err
+	}
+	// no recipients is not an error...
+	if len(pkcs11Recipients) == 0 {
+		return nil, nil
+	}
+
+	jsonString, err := pkcs11.EncryptMultiple(pkcs11Recipients, optsData)
+	if err != nil {
+		return nil, errors.Wrapf(err, "PKCS11 EncryptMulitple failed")
+	}
+	return jsonString, nil
+}
+
+func (kw *pkcs11KeyWrapper) UnwrapKey(dc *config.DecryptConfig, jsonString []byte) ([]byte, error) {
+	var pkcs11PrivKeys []*pkcs11uri.Pkcs11URI
+
+	privKeys := kw.GetPrivateKeys(dc.Parameters)
+	if len(privKeys) == 0 {
+		return nil, errors.New("No private keys found for PKCS11 decryption")
+	}
+	for _, privKey := range privKeys {
+		key, err := utils.ParsePrivateKey(privKey, nil, "PKCS11")
+		if err != nil {
+			return nil, err
+		}
+		switch pkcs11Key := key.(type) {
+		case *pkcs11.Pkcs11KeyFileObject:
+			if dc.Pkcs11Config != nil {
+				pkcs11Key.Uri.SetModuleDirectories(dc.Pkcs11Config.ModuleDirectories)
+			}
+			pkcs11PrivKeys = append(pkcs11PrivKeys, pkcs11Key.Uri)
+		default:
+			continue
+		}
+	}
+
+	plaintext, err := pkcs11.Decrypt(pkcs11PrivKeys, jsonString)
+	if err == nil {
+		return plaintext, nil
+	}
+
+	return nil, errors.New("PKCS11: No suitable private key found for decryption")
+}
+
+func (kw *pkcs11KeyWrapper) NoPossibleKeys(dcparameters map[string][][]byte) bool {
+	return len(kw.GetPrivateKeys(dcparameters)) == 0
+}
+
+func (kw *pkcs11KeyWrapper) GetPrivateKeys(dcparameters map[string][][]byte) [][]byte {
+	return dcparameters["pkcs11-yamls"]
+}
+
+func (kw *pkcs11KeyWrapper) GetKeyIdsFromPacket(unused string) ([]uint64, error) {
+	return nil, nil
+}
+
+func (kw *pkcs11KeyWrapper) GetRecipients(unused string) ([]string, error) {
+	return []string{"[pkcs11]"}, nil
+}
+
+func addPubKeys(dc *config.DecryptConfig, pubKeys [][]byte) ([]interface{}, error) {
+	var pkcs11Keys []interface{}
+	if len(pubKeys) == 0 {
+		return pkcs11Keys, nil
+	}
+
+	for _, pubKey := range pubKeys {
+		key, err := utils.ParsePublicKey(pubKey, "PKCS11")
+		if err != nil {
+			return nil, err
+		}
+		switch p11Uri := key.(type) {
+		case *pkcs11uri.Pkcs11URI:
+			if dc.Pkcs11Config != nil {
+				p11Uri.SetModuleDirectories(dc.Pkcs11Config.ModuleDirectories)
+			}
+		}
```

Comment:
Check default case

Reply:
Sorry, which default case?

Reply:
for type casting, in case it runs into a casting error. or something like `p11Uri, err := key.(*pkcs11uri.Pkcs11URI)`

Reply:
What casting error? Here's the surrounding code:

````
		switch p11Uri := key.(type) {
		case *pkcs11uri.Pkcs11URI:
			if dc.Pkcs11Config != nil {
				p11Uri.SetModuleDirectories(dc.Pkcs11Config.ModuleDirectories)
			}
		}